pkg/inject/inject.go

```diff
@@ -384,15 +384,56 @@ func (conf *ResourceConfig) CreateDefaultOpaquePortsPatch() ([]byte, error) {
 			return nil, err
 		}
 	} else if !conf.HasWorkloadAnnotation(k8s.ProxyOpaquePortsAnnotation) {
-		opaquePorts := conf.GetValues().Proxy.OpaquePorts
-		patch, err = conf.CreateAnnotationPatch(opaquePorts)
+		defaultPorts := strings.Split(conf.GetValues().Proxy.OpaquePorts, ",")
+		var filteredPorts []string
+		// If the workload is a pod, then only add the default opaque ports
+		// that it exposes as container ports.
+		//
+		// If the workload is a service, then only add the default opaque
+		// ports that are exposed as a service port, or targeted as a
+		// targetPort.
+		if conf.IsPod() {
+			filteredPorts = conf.FilterPodOpaquePorts(defaultPorts)
+		} else if conf.IsService() {
+			service := conf.workload.obj.(*corev1.Service)
+			for _, p := range service.Spec.Ports {
+				port := strconv.Itoa(int(p.Port))
+				if util.ContainsString(port, defaultPorts) {
```

Comment:
should this case only apply if targetPort is not set?  i.e. if the port simply maps to itself as the target port.

Reply:
Ok 3931fd9 addresses this and adds some additional tests to make sure we don't set a port as opaque if it targets a port that is _not_ opaque.